tajo-core/src/main/java/org/apache/tajo/querymaster/Stage.java

```diff
@@ -1300,6 +1301,52 @@ protected void stopFinalization() {
     stopShuffleReceiver.set(true);
   }
 
+  private void finalizeShuffleReport(StageShuffleReportEvent event, ShuffleType type) {
+    if(!checkIfNeedFinalizing(type)) return;
+
+    TajoWorkerProtocol.ExecutionBlockReport report = event.getReport();
+
+    if (!report.getReportSuccess()) {
+      stopFinalization();
+      LOG.error(getId() + ", " + type + " report are failed. Caused by:" + report.getReportErrorMessage());
+      eventHandler.handle(new StageEvent(getId(), StageEventType.SQ_FAILED));
+    }
+
+    completedShuffleTasks.addAndGet(report.getSucceededTasks());
+    if (report.getIntermediateEntriesCount() > 0) {
+      for (IntermediateEntryProto eachInterm : report.getIntermediateEntriesList()) {
+        hashShuffleIntermediateEntries.add(new IntermediateEntry(eachInterm));
+      }
+    }
+
+    if (completedShuffleTasks.get() >= succeededObjectCount) {
+      LOG.info(getId() + ", Finalized " + type + " reports: " + completedShuffleTasks.get());
+      eventHandler.handle(new StageEvent(getId(), StageEventType.SQ_STAGE_COMPLETED));
+      if (timeoutChecker != null) {
+        stopFinalization();
+        synchronized (timeoutChecker){
+          timeoutChecker.notifyAll();
+        }
+      }
+    } else {
+      LOG.info(getId() + ", Received " + type + " reports " +
+          completedShuffleTasks.get() + "/" + succeededObjectCount);
+    }
+  }
+
+  /**
+   * HASH_SHUFFLE, SCATTERED_HASH_SHUFFLE should get report from worker nodes when ExecutionBlock is stopping.
```

Comment:
It would be great if you add a comment that describes why we don't need to collect reports when the shuffle type is RANGE_SHUFFLE.